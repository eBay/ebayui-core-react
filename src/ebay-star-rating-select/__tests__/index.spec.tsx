import React from 'react'
import { render, fireEvent, cleanup } from '@testing-library/react';
import { Isolated,Fieldset } from './mocks'

const htmlSnap =(ui: React.ReactElement): void => {
    const { asFragment } = render(ui);
    expect(asFragment()).toMatchSnapshot();
}

let component;
let onChangeSpy = jest.fn()
let onFocusSpy = jest.fn()
let onKeyDownSpy = jest.fn()

describe('star-rating-select', () => {
    it('renders defaults', async () => {
        await htmlSnap(<Isolated a11yText={null} a11yStarText={null}/>);
    });

    it('renders basic fieldset', async () => {
        await htmlSnap(<Fieldset/>);
    });

    it('renders with 0 selected', async () => {
        await htmlSnap(<Isolated value="0"/>);
    });

    it('renders with 2 selected', async () => {
        await htmlSnap(<Isolated value="2"/>);
    });

    it('renders with 5 selected', async () => {
        await htmlSnap(<Isolated value="5"/>);
    });

    describe('given star rating', () => {
        afterEach(cleanup);

        beforeEach(async () => {
            onChangeSpy.mockReset()
            component = await render(<Isolated onChange={onChangeSpy} />);
            await fireEvent.click(component.getByLabelText('2 stars'));
        });

        it('then it emits the event', () => {
            const changeEvents = onChangeSpy.mock.calls;
            expect(changeEvents.length).toBe(1);

            const eventArgs = changeEvents[0];
            expect(eventArgs.length).toBe(2);
            expect(eventArgs[1]).toBe(2);
        });

        describe('when star is clicked', () => {
            beforeEach(async () => {
                await fireEvent.click(component.getByLabelText('4 stars'));
            });

            it('then it emits the event', () => {
                const changeEvents = onChangeSpy.mock.calls;
                expect(changeEvents.length).toBe(2);

                const eventArgs = changeEvents[1];
                expect(eventArgs.length).toBe(2);
                expect(eventArgs[1]).toBe(4);
            });
        });
    });

    describe('given star is disabled', () => {
        beforeEach(async () => {
            onChangeSpy.mockReset()
            component = await render(<Isolated onChange={onChangeSpy} disabled/>);
        });

        describe('when star is clicked', () => {
            beforeEach(async () => {
                await fireEvent.click(component.getByLabelText('2 stars'));
            });

            it("then it doesn't emit the event", () => {
                const changeEvents = onChangeSpy.mock.calls;
                expect(changeEvents.length).toBe(0);
            });
        });
    });

    describe('when native focus event is fired on star', () => {
        beforeEach(async () => {
            onFocusSpy.mockReset()
            component = await render(<Isolated onFocus={onFocusSpy}  />);
            await fireEvent.focus(component.getByLabelText('2 stars'));
        });

        it('then it emits the event', () => {
            const events = onFocusSpy.mock.calls;
            expect(events.length).toBe(1);

            const [[eventArg,value]] = events;
            expect(value).toBe(2);
        });
    });

    describe('when native keyboard event is fired on star', () => {
        beforeEach(async () => {
            onKeyDownSpy.mockReset()
            component = await render(<Isolated onKeyDown={onKeyDownSpy}/>);
            await fireEvent.keyDown(component.getByLabelText('5 stars'));
        });

        it('then it emits the event', () => {
            const events = onKeyDownSpy.mock.calls;
            expect(events.length).toBe(1);
            const [[eventArg,value]] = events;
            expect(value).toBe(5);
        });
    });
});
