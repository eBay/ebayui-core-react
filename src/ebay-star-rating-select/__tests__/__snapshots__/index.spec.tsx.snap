// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`star-rating-select renders basic fieldset 1`] = `
<DocumentFragment>
  <fieldset>
    <legend>
      Rate Product
    </legend>
    <div
      class="star-rating-select"
    >
      <span
        class="star-rating-select__radio"
      >
        <input
          aria-label="1 star"
          class="star-rating-select__control"
          name="star-rating"
          type="radio"
          value="1"
        />
        <span
          class="star-rating-select__radio-icon"
        >
          <svg
            aria-hidden="true"
            class="star-rating__icon icon icon--star-dynamic"
            focusable="false"
            height="24px"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="#icon-star-dynamic"
            />
          </svg>
        </span>
      </span>
      <span
        class="star-rating-select__radio"
      >
        <input
          aria-label="2 stars"
          class="star-rating-select__control"
          name="star-rating"
          type="radio"
          value="2"
        />
        <span
          class="star-rating-select__radio-icon"
        >
          <svg
            aria-hidden="true"
            class="star-rating__icon icon icon--star-dynamic"
            focusable="false"
            height="24px"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="#icon-star-dynamic"
            />
          </svg>
        </span>
      </span>
      <span
        class="star-rating-select__radio"
      >
        <input
          aria-label="3 stars"
          class="star-rating-select__control"
          name="star-rating"
          type="radio"
          value="3"
        />
        <span
          class="star-rating-select__radio-icon"
        >
          <svg
            aria-hidden="true"
            class="star-rating__icon icon icon--star-dynamic"
            focusable="false"
            height="24px"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="#icon-star-dynamic"
            />
          </svg>
        </span>
      </span>
      <span
        class="star-rating-select__radio"
      >
        <input
          aria-label="4 stars"
          class="star-rating-select__control"
          name="star-rating"
          type="radio"
          value="4"
        />
        <span
          class="star-rating-select__radio-icon"
        >
          <svg
            aria-hidden="true"
            class="star-rating__icon icon icon--star-dynamic"
            focusable="false"
            height="24px"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="#icon-star-dynamic"
            />
          </svg>
        </span>
      </span>
      <span
        class="star-rating-select__radio"
      >
        <input
          aria-label="5 stars"
          class="star-rating-select__control"
          name="star-rating"
          type="radio"
          value="5"
        />
        <span
          class="star-rating-select__radio-icon"
        >
          <svg
            aria-hidden="true"
            class="star-rating__icon icon icon--star-dynamic"
            focusable="false"
            height="24px"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="#icon-star-dynamic"
            />
          </svg>
        </span>
      </span>
    </div>
  </fieldset>
</DocumentFragment>
`;

exports[`star-rating-select renders defaults 1`] = `
<DocumentFragment>
  <div
    class="star-rating-select"
  >
    <span
      class="star-rating-select__radio"
    >
      <input
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="1"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="2"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="3"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="4"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="5"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`star-rating-select renders with 0 selected 1`] = `
<DocumentFragment>
  <div
    aria-label="Rate product"
    class="star-rating-select"
    role="radiogroup"
  >
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="1 star"
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="1"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="2 stars"
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="2"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="3 stars"
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="3"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="4 stars"
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="4"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="5 stars"
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="5"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`star-rating-select renders with 2 selected 1`] = `
<DocumentFragment>
  <div
    aria-label="Rate product"
    class="star-rating-select"
    role="radiogroup"
  >
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="1 star"
        class="star-rating-select__control star-rating-select__control--filled"
        name="star-rating"
        type="radio"
        value="1"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="2 stars"
        checked=""
        class="star-rating-select__control star-rating-select__control--filled"
        name="star-rating"
        type="radio"
        value="2"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="3 stars"
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="3"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="4 stars"
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="4"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="5 stars"
        class="star-rating-select__control"
        name="star-rating"
        type="radio"
        value="5"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`star-rating-select renders with 5 selected 1`] = `
<DocumentFragment>
  <div
    aria-label="Rate product"
    class="star-rating-select"
    role="radiogroup"
  >
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="1 star"
        class="star-rating-select__control star-rating-select__control--filled"
        name="star-rating"
        type="radio"
        value="1"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="2 stars"
        class="star-rating-select__control star-rating-select__control--filled"
        name="star-rating"
        type="radio"
        value="2"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="3 stars"
        class="star-rating-select__control star-rating-select__control--filled"
        name="star-rating"
        type="radio"
        value="3"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="4 stars"
        class="star-rating-select__control star-rating-select__control--filled"
        name="star-rating"
        type="radio"
        value="4"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
    <span
      class="star-rating-select__radio"
    >
      <input
        aria-label="5 stars"
        checked=""
        class="star-rating-select__control star-rating-select__control--filled"
        name="star-rating"
        type="radio"
        value="5"
      />
      <span
        class="star-rating-select__radio-icon"
      >
        <svg
          aria-hidden="true"
          class="star-rating__icon icon icon--star-dynamic"
          focusable="false"
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-star-dynamic"
          />
        </svg>
      </span>
    </span>
  </div>
</DocumentFragment>
`;
